mesonbuild/build.py

```diff
@@ -1591,11 +1591,16 @@ def check_module_linking(self):
             if isinstance(link_target, SharedModule):
                 if self.environment.machines[self.for_machine].is_darwin():
                     raise MesonException(
-                        'target links against shared modules. This is not permitted on OSX')
+                        f'target {self.name} links against shared module {link_target.name}. This is not permitted on OSX')
                 else:
-                    mlog.warning('target links against shared modules. This '
-                                 'is not recommended as it is not supported on some '
-                                 'platforms')
+                    mlog.deprecation(f'target {self.name} links against shared module {link_target.name}, which is incorrect.'
+                            '\n             '
+                            f'This will be an error in the future, so please use shared_library() for {link_target.name} instead.'
+                            '\n             '
+                            f'If shared_module() was used for {link_target.name} because it has references to undefined symbols,'
+                            '\n             '
+                            'use shared_libary() with `override_options: [\'b_lundef=false\']` instead.')
+                    link_target.backwards_compat_want_soname = True
                 return
```

Comment:
This return statement causes the workaround to fail if a target is linked against multiple shared_module targets. I think it should be removed.

Reply:
As an example, see gthumb.
https://bugs.gentoo.org/827143
https://gitlab.gnome.org/GNOME/gthumb/-/blob/3.12.0/extensions/photo_importer/meson.build#L23

Reply:
https://github.com/mesonbuild/meson/pull/9640